src/uct/self/self_ep.c

```diff
@@ -23,19 +23,81 @@ static UCS_CLASS_CLEANUP_FUNC(uct_self_ep_t)
 UCS_CLASS_DEFINE_DELETE_FUNC(uct_self_ep_t, uct_ep_t);
 
+static void
+uct_self_ep_am_reserve_and_get_new_buffer(ucs_status_t *status,
+                                          uct_self_iface_t *self_iface,
+                                          const uct_self_msg_desc_t *desc)
+{
+    /* Reserve the buffer */
+    uct_recv_desc_iface(desc->am_recv_data) = &self_iface->super.super;
+
+    /* get new buffer from memory pool
+    * No need to copy data from payload to desc->am_recv_data */
+    self_iface->msg_cur_desc = ucs_mpool_get(&self_iface->msg_desc_mp);
+    VALGRIND_MAKE_MEM_DEFINED(self_iface->msg_cur_desc,
+                              sizeof(*(self_iface->msg_cur_desc)));
+    if (NULL == self_iface->msg_cur_desc) {
+        ucs_error("Failed to get next descriptor in SELF MP storage");
+        *status = UCS_ERR_NO_RESOURCE;
+    }
+}
+
 ucs_status_t uct_self_ep_am_short(uct_ep_h ep, uint8_t id, uint64_t header,
                                   const void *payload, unsigned length)
 {
-    uct_base_iface_t *local_iface = 0;
+    ucs_status_t status;
+    uct_self_iface_t *self_iface = 0;
+    uct_self_msg_desc_t *desc = 0;
 
-    ucs_trace_data("TX: self AM [%d] buf=%p len=%u", id, payload, length);
-    local_iface = ucs_derived_of(ep->iface, uct_base_iface_t);
-    /* stub for testing */
-    uct_iface_invoke_am(local_iface, id, (void *)payload, length, (void *)payload);
-    return UCS_OK;
+    self_iface = ucs_derived_of(ep->iface, uct_self_iface_t);
+
+    UCT_CHECK_AM_ID(id);
+    uct_iface_trace_am(&self_iface->super, UCT_AM_TRACE_TYPE_SEND, id,
+                       payload, length + sizeof(header), "TX: AM_SHORT");
+    UCT_CHECK_LENGTH(length + sizeof(header),
+                     self_iface->data_length - sizeof(uct_self_msg_desc_t),
+                     "am_short");
+
+    desc = self_iface->msg_cur_desc;
+    status = uct_iface_invoke_am(&self_iface->super, id, (void *)payload, length,
+                                 (void *)desc->am_recv_data);
+    if (ucs_unlikely(UCS_OK != status)) {
+        uct_self_ep_am_reserve_and_get_new_buffer(&status, self_iface, desc);
+    }
+
+    return status;
+}
+
+ssize_t uct_self_ep_am_bcopy(uct_ep_h tl_ep, uint8_t id,
+                             uct_pack_callback_t pack_cb, void *arg)
+{
+    ucs_status_t status;
+    uct_self_iface_t *self_iface = 0;
+    uct_self_msg_desc_t *desc = 0;
+    size_t length = 0;
+
+    self_iface = ucs_derived_of(tl_ep->iface, uct_self_iface_t);
+
+    desc = self_iface->msg_cur_desc;
+    length = pack_cb(desc->am_recv_data, arg);
+
+    UCT_CHECK_AM_ID(id);
+    uct_iface_trace_am(&self_iface->super, UCT_AM_TRACE_TYPE_SEND, id,
+                       arg, length, "TX: AM_BCOPY");
+    UCT_CHECK_LENGTH(length, self_iface->data_length - sizeof(uct_self_msg_desc_t),
+                     "am_bcopy");
+
+    status = uct_iface_invoke_am(&self_iface->super, id, (void *)desc->am_recv_data,
```

Comment:
@yosefe The main idea is make uct_iface_invoke_am call without additional branches (no guarantee that it will be inlined from uct_self_ep_am_reserve_and_get_new_buffer). The status handling (if not UCS_OK) marked as ucs_unlikly in hope this help to gcc optimize the hot path better.
If you think it is better to move "invoke_am" to the common function, of course, I will do.